internal/server/recon/place_name.go

```diff
@@ -0,0 +1,271 @@
+	"github.com/datacommonsorg/mixer/internal/server/v1/propertyvalues"
+	"github.com/datacommonsorg/mixer/internal/store"
+	"github.com/datacommonsorg/mixer/internal/util"
+	"golang.org/x/sync/errgroup"
+	"googlemaps.github.io/maps"
+)
+
+const (
+	maxNumPlacesPerRequest    = 5000
+	maxMapsAPICallsInParallel = 25
+)
+
+// ResolvePlaceNames implements API for Recon.ResolvePlaceNames.
+func ResolvePlaceNames(
+	ctx context.Context,
+	in *pb.ResolvePlaceNamesRequest,
+	store *store.Store,
+	mapsClient *maps.Client,
+) (*pb.ResolvePlaceNamesResponse, error) {
+	if l := len(in.GetPlaces()); l == 0 {
+		return nil, fmt.Errorf("empty input")
+	} else if l > maxNumPlacesPerRequest {
+		return nil, fmt.Errorf(
+			"exceeded max number of places per request (%d): %d",
+			maxNumPlacesPerRequest, l)
+	}
+
+	type placeInfo struct {
+		placeName string
+		placeType string
+	}
+
+	// Load input.
+	nameSet := map[string]struct{}{}
+	placeInfoSet := map[placeInfo]struct{}{}
+	for _, place := range in.GetPlaces() {
+		name := place.GetName()
+		if name == "" {
+			return nil, fmt.Errorf("empty place name")
+		}
+		nameSet[name] = struct{}{}
+		placeInfoSet[placeInfo{name, place.GetType()}] = struct{}{}
+	}
+
+	// Get place IDs.
+	nameToPlaceIDs, placeIDSet, err := resolvePlaceIDs(ctx, mapsClient, nameSet)
+	if err != nil {
+		return nil, err
+	}
+
+	// Resolve place IDs to get DCIDs.
+	placeIDToDCIDs, dcidSet, err := resolveDCIDs(ctx, placeIDSet, store)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get types of the DCIDs.
+	dcidToTypeSet, err := getPlaceTypes(ctx, dcidSet, store)
+	if err != nil {
+		return nil, err
+	}
+
+	// Assemble results.
+	resp := &pb.ResolvePlaceNamesResponse{}
+	for info := range placeInfoSet {
+		place := &pb.ResolvePlaceNamesResponse_Place{
+			Name: info.placeName,
+			Type: info.placeType,
+		}
+
+		if placeIDs, ok := nameToPlaceIDs.Load(info.placeName); ok {
+			allDCIDs := []string{}
+			for _, placeID := range placeIDs.([]string) {
+				if dcids, ok := placeIDToDCIDs[placeID]; ok {
+					allDCIDs = append(allDCIDs, dcids...)
+				}
+			}
+			sort.Strings(allDCIDs)
+
+			if len(allDCIDs) != 0 {
+				if info.placeType == "" {
+					// No type filtering.
+					place.Dcids = allDCIDs
+				} else {
+					// Type filtering.
+					filteredDCIDs := []string{}
+					for _, dcid := range allDCIDs {
+						typeSet, ok := dcidToTypeSet[dcid]
+						if !ok {
+							continue
+						}
+						if _, ok := typeSet[info.placeType]; ok {
+							filteredDCIDs = append(filteredDCIDs, dcid)
+						}
+					}
+					place.Dcids = filteredDCIDs
+				}
+			}
+		}
+
+		resp.Places = append(resp.Places, place)
+	}
+
+	return resp, nil
+}
+
+func resolvePlaceIDs(
+	ctx context.Context,
+	mapsClient *maps.Client,
+	nameSet map[string]struct{},
+) (*sync.Map /* name -> place IDs */, *sync.Map /* place ID set */, error) {
+	type placeInfo struct {
+		name     string
+		placeIDs []string
+	}
+	var wg sync.WaitGroup
+
+	// Send place names to channel.
```

Comment:
I confess I only vaguely follow go synchronization.... wonder if it would help to provide a preface to how we're parallelizing?

My understanding....  Set up a "request" channel of names to be resolved, set up `maxMapsAPICallsInParallel` "response" channels for resolved place IDs, run `maxMapsAPICallsInParallel` goroutines that each reads from the "request" channel, calls place API and writes to the "response" channel ?